stdlib/public/core/Collection.swift

```diff
@@ -260,9 +260,10 @@ extension CollectionType {
 
   /// Returns the first element of `self`, or `nil` if `self` is empty.
   ///
-  /// - Complexity: O(1)
+  /// - Complexity: O(1) in most cases, possibly O(N) for lazy collections.
   public var first: Generator.Element? {
-    return isEmpty ? nil : self[startIndex]
+    var gen = generate()
+    return gen.next()
```

Comment:
The problem is that the generator approach is inherently more expensive in some cases.  For example, consider `(0..<20).lazy.map { veryExpensiveCall($0) }.isEmpty`.  I think comparing indices is the obvious way to go.